### PR TITLE
Aric types v4

### PR DESCRIPTION
**Pull request by lasryaric**
# What's included

- New shared `@dust-tt/types` npm package (`/types/`)
- All the infra to dev (`cd types/ && npm install &&  npm start`)
- All the infra to deploy
- One Typescript type use case: `KeyType`
- One io-ts type use case: `PostContentFragmentRequestBodySchema`

# Deployment

We need to deploy and observe success for all apps relying on the `k8s/cloud-build.sh` script.
They are listed here in order of increasing sensitivity.

- [x] Lock deployment on the #deployments channel on Slack.
- [x] Deploy `alerting/temporal` and observe success
- [x] Deploy `blog` and observe success
- [x] Deploy `docs` and observe success
- [x] Deploy `front-edge` and observe success
- [x] Deploy `connectors-edge` and observe success 
- [x] Deploy `connectors` and observe success 
- [x] Deploy `front` and observe success 
- [x] Deploy `core` and observe success 
- [ ] Update the dev runbook

# Proposal

Created a `@dust-tt/types` package that lives at the root of the dust directory.
The package is imported by `connectors/` and `front/` via relative path import:

```
# front/package.json OR connectors/package.json
"dependencies":{
"@dust-tt/types":"../types"
} 
```

This makes deploy time and build time straightforward, with the following actions to execute:
In dev mode:
```
 cd types/ && npm install && npm start 
 Watching and Building...
```

In production:
```
front/Dockerfile
RUN cd /types/ && npm install && npm run build

WORKDIR /front
npm install
....
```

# Other solutions explored

## npm workspaces

NPM workspaces looked like what we need on paper but there is one main caveat: there would be only one `package-lock.json` per workspace, which means one lock file for all our apps (front/, connectors/, types/, and eventually sparkle).
This works well for developing a list of packages that you need to bundle and npm publisbh, but is not suited to deploying packages as Docker apps. If you want to know more about that, you can read that NPM github issue:
https://github.com/npm/rfcs/issues/554

## npm publish (or tarball) + npm link
We could also simply import `@dust-tt/types` like any other package, and manage a publishing / installing version for each deployment, either manually or via an automated CI, but this introduces a lot of moving parts at deployment, which made us rule out this solution.

For that to work in development mode, one would need to execute the following commands:

```
cd /types/
npm link
cd /front/
npm link types
```

# Other solutions?

I am not aware of any other solution, but if you know one, happy to explore it.

### CODE REVIEW

**Review by spolu — COMMENTED** [review]
Overall approach looks good.  Left a few questions

**Comment by lasryaric** [comment]
@spolu I did not mention it but this is POC for front. IF we agree on the approach design, I will make it work for all services.
I will clean up all the non needed stuff added by the package creator too.

File: front/Dockerfile [hunk 1]
```diff
@@ -1,14 +1,22 @@
 FROM node:18.15.0 as front
 
+RUN mkdir -p /app
```
[the remark this answers is not in the document]
Comment (lasryaric) [reply front/Dockerfile:3]: Not be required, wanted it to be explicit.
If you start the image `node:18/15.5`, it has no /app directory. I find it more confusing than anything else to not see it explicitly created.

File: front/Dockerfile [hunk 2]
```diff
@@ -1,14 +1,22 @@
 FROM node:18.15.0 as front
 
+RUN mkdir -p /app 
+RUN mkdir -p /types
+
+WORKDIR /types
+COPY /types/ /types
```
[the remark this answers is not in the document]
Comment (lasryaric) [reply front/Dockerfile:7]: ok I will change it. I like explicit things in script like that, but you have the WORKDIR juste before I agree.

File: types/package.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  "version": "0.1.0",
+  "author": "Dust",
+  "main": "dist/index.js",
+  "module": "dist/types.esm.js",
+  "typings": "dist/index.d.ts",
+  "files": [
+    "dist",
+    "src"
+  ],
+  "scripts": {
```
[the remark this answers is not in the document]
Comment (lasryaric) [reply types/package.json:12]: we need start, not test

File: types/package.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    "analyze": "size-limit --why",
+    "build": "dts build",
+    "lint": "dts lint",
+    "format": "dts lint --fix",
+    "prepare": "dts build",
+    "size": "size-limit",
+    "start": "dts watch",
+    "test": "dts test"
+  },
+  "husky": {
```
[the remark this answers is not in the document]
Comment (lasryaric) [reply types/package.json:22]: Will remove. It's to manage git hook. Was added by dts-cli, which helped me create the /types/ package.

File: types/package.json [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    "singleQuote": true,
+    "trailingComma": "es5"
+  },
+  "jest": {
+    "testEnvironment": "node"
+  },
+  "engines": {
+    "node": ">=12"
+  },
+  "size-limit": [
```
[the remark this answers is not in the document]
Comment (lasryaric) [reply types/package.json:39]: no

File: k8s/cloud-build.sh [hunk 6]
```diff
@@ -15,4 +14,4 @@ echo "Current working directory is $(pwd)"
 
 # Start the build and get its ID
 echo "Starting build..."
-BUILD_ID=$(gcloud builds submit --quiet --config "${SCRIPT_DIR}/cloudbuild.yaml" --substitutions=SHORT_SHA=$(git rev-parse --short HEAD),_IMAGE_NAME=$1 --format='value(id)' .)
+BUILD_ID=$(gcloud builds submit --quiet --config "${SCRIPT_DIR}/cloudbuild.yaml" --substitutions=SHORT_SHA=$(git rev-parse --short HEAD),_IMAGE_NAME=$1 --format='value(id)' ../)
```
Comment (fontanierh) [inline k8s/cloud-build.sh:17]: Won't this break docker images that you didn't change?

File: k8s/cloudbuild.yaml [hunk 7]
```diff
@@ -9,6 +9,8 @@ steps:
       - -t
       - gcr.io/$PROJECT_ID/${_IMAGE_NAME}-image:latest
       - --push
+      - -f
+      - ./front/Dockerfile
```
[the remark this answers is not in the document]
Comment (fontanierh) [reply k8s/cloudbuild.yaml:13]: Yep we def don't want `front` here

**Review by fontanierh — COMMENTED** [review]
directionally not against it, but have you tested HMR on `front` when you update code in `types` ? I have the feeling that it's not going to work as expected (which IMO is key here)

**Comment by lasryaric** [comment]
Putting this PR back to draft mode. The goal was to agree on the overall design, but now I need to make it work for all packages and do some clean up.

**Comment by lasryaric** [comment]
(No need to review until I ping you back)

File: .github/workflows/build-and-lint-connectors.yml [hunk 8]
```diff
@@ -16,5 +16,7 @@ jobs:
           node-version: 18.15.0
           cache: "npm"
           cache-dependency-path: ./connectors/package-lock.json
+      - working-directory: types
```
Comment (lasryaric) [inline .github/workflows/build-and-lint-connectors.yml:19]: We need a build-and-lint types/ and only call npm install here.

File: .github/workflows/build-and-lint-front.yml [hunk 9]
```diff
@@ -16,5 +16,7 @@ jobs:
           node-version: 18.15.0
           cache: "npm"
           cache-dependency-path: ./front/package-lock.json
+      - working-directory: types
+        run: npm install && npm run lint &&  npm run build
```
Comment (spolu) [inline .github/workflows/build-and-lint-front.yml:20]: can we skip some of it to keep front fast? I presume lint for sure? Maybe even build?

File: .github/workflows/deploy-front.yml [hunk 10]
```diff
@@ -41,7 +41,7 @@ jobs:
       - name: Build the image on Cloud Build
         run: |
           chmod +x ./k8s/cloud-build.sh
-          ./k8s/cloud-build.sh front
+          ./k8s/cloud-build.sh front ./front/Dockerfile ../
```
Comment (spolu) [inline .github/workflows/deploy-front.yml:44]: Why `../`?

File: k8s/cloud-build.sh [hunk 11]
```diff
@@ -1,11 +1,12 @@
...
 # first arg : name of image to build
-# second arg (optional) : working directory. If not provided, it defaults to "../$1"
+# second arg: path to the Dockerfile, relative to the resolved WORKING_DIR.
+# thid arg (optional) : working directory. If not provided, it defaults to "../$1"
 set -e
 
-
 SCRIPT_DIR="$(cd "$(dirname "$0")" >/dev/null 2>&1 && pwd)"
-WORKING_DIR="${2:-${SCRIPT_DIR}/../$1}"
+WORKING_DIR="${3:-${SCRIPT_DIR}/../$1}"
```
Comment (spolu) [inline k8s/cloud-build.sh:8]: Why not just keep the new $3 ?

**Comment by spolu** [comment]
Looks great! Left a few questions on things I felt unsure about :+1:

**Review by fontanierh — COMMENTED** [review]
The code LGTM, super bullish and thanks for doing it 🙏 

A bit scary change since it affects dev workflow and deployments, so I'm going to spend a little time testing it locally before approving.

**Review by lasryaric — COMMENTED** [review]
@spolu answered your comments.

File: .github/workflows/deploy-front.yml (hunk 10 above)
Comment (spolu) [reply .github/workflows/deploy-front.yml:44]: Right but for `connectors`, `connectors-edge`, `front-edge` it's `./k8s/cloud-build.sh front ./front/Dockerfile ./`

File: k8s/cloud-build.sh (hunk 11 above)
Comment (spolu) [reply k8s/cloud-build.sh:8]: I would be in favor of removing the default so that it fails faster

File: .github/workflows/deploy-front.yml (hunk 10 above)
Comment (lasryaric) [reply .github/workflows/deploy-front.yml]: It's an error on my side. I was not able to test this one yet. Fixed.

**Review by lasryaric — COMMENTED** [review]
Addressed @spolu's comments.

**Review by fontanierh — APPROVED** [review]
I think it's good, it looks like it works.
Awesome ! 🙏

**Pull request merged**